src/mono/wasm/runtime/binding_support.js

```diff
@@ -1752,6 +1779,238 @@ var BindingSupportLib = {
+		_mono_wasm_web_socket_send_buffering: function (ws, message_ptr, length, message_type, end_of_message) {
+			var buffer = ws[this.wasm_ws_pending_send_buffer];
+			var offset = 0;
+
+			if (buffer) {
+				offset = ws[this.wasm_ws_pending_send_buffer_offset];
+				// match desktop WebSocket behavior by copying message_type of the first part
+				message_type = ws[this.wasm_ws_pending_send_buffer_type];
+				// if not empty message, append to existing buffer
+				if (length !== 0) {
+					const view = Module.HEAPU8.subarray(message_ptr, message_ptr + length);
+					if (offset + length > buffer.length) {
+						const newbuffer = new Uint8Array((offset + length + 50) * 1.5); // exponential growth
+						newbuffer.set(buffer, 0, offset);// copy previous buffer
+						newbuffer.set(view, offset);// append copy at the end
+						ws[this.wasm_ws_pending_send_buffer] = buffer = newbuffer;
+					}
+					else {
+						buffer.set(view, offset);// append copy at the end
+					}
+					offset += length;
+					ws[this.wasm_ws_pending_send_buffer_offset] = offset;
+				}
+			}
+			else if (!end_of_message) {
+				// create new buffer
+				if (length !== 0) {
+					const view = Module.HEAPU8.subarray(message_ptr, message_ptr + length + 1024);
+					buffer = new Uint8Array(view); // copy
+					offset = length;
+					ws[this.wasm_ws_pending_send_buffer_offset] = offset;
+					ws[this.wasm_ws_pending_send_buffer] = buffer;
+				}
+				ws[this.wasm_ws_pending_send_buffer_type] = message_type;
+			}
+			else {
+				// use the buffer only localy
+				if (length !== 0) {
+					const memoryView = Module.HEAPU8.subarray(message_ptr, message_ptr + length);
+					buffer = memoryView; // send will make a copy
+					offset = length;
+				}
+			}
+			// buffer was updated, do we need to trim and convert it to final format ?
+			if (end_of_message) {
+				if (offset == 0) {
+					return new Uint8Array();
+				}
+				if (message_type === 0) {
+					// text, convert from UTF-8 bytes to string, because of bad browser API
+					if (this._text_decoder_utf8 === undefined) {
+						this._text_decoder_utf8 = new TextDecoder('utf-8');
+					}
+
+					// See https://github.com/whatwg/encoding/issues/172
+					var bytes = typeof SharedArrayBuffer !== 'undefined' && buffer instanceof SharedArrayBuffer
+						? buffer.slice(0, offset)
+						: buffer.subarray(0, offset);
+
+					return this._text_decoder_utf8.decode(bytes);
+				} else {
+					// binary, view to used part of the buffer
+					return buffer.subarray(0, offset);
+				}
+			}
+			return null;
+		},
+		_mono_wasm_web_socket_send_and_wait: function (ws, buffer, thenable_js_handle) {
+			// send and return promise
+			ws.send(buffer);
+			ws[this.wasm_ws_pending_send_buffer] = null;
+
+			// if the send buffer is small, we don't block so that the throughput doesn't suffer. 
+			// Otherwise we block so that we apply some backpresure to the application sending large data.
+			if (ws.bufferedAmount < 65536) {
```

Comment:
We might want to make this a named constant near the top of the file, if not a publicly exposed value that can be changed.